File: lib/features/promotions/models/promotion.dart
class Promotion {
  const Promotion({
    required this.id,
    required this.type,
    required this.title,
    required this.description,
    required this.store,
    required this.discountText,
    required this.landingUrl,
    required this.featured,
    required this.updatedAt,
    this.code,
    this.affiliateUrl,
    this.imageUrl,
    this.providerId,
    this.monetizationMode = 'affiliate',
    this.availabilityCountries = const <String>[],
    this.isGlobal = false,
    this.validFrom,
    this.validUntil,
  });

  final String id;
  final String type;
  final String title;
  final String description;
  final String store;
  final String discountText;
  final String? code;
  final String landingUrl;
  final String? affiliateUrl;
  final String? imageUrl;
  final String? providerId;
  final String monetizationMode;
  final List<String> availabilityCountries;
  final bool isGlobal;
  final DateTime? validFrom;
  final DateTime? validUntil;
  final bool featured;
  final DateTime updatedAt;

  bool get hasCode => code != null && code!.trim().isNotEmpty;

  bool get isFlashSale => type == 'flash_sale';

  bool get hasDeadline => validUntil != null;

  factory Promotion.fromJson(Map<String, dynamic> json) {
    return Promotion(
      id: _string(json['id']),
      type: _string(json['type'], fallback: 'sale'),
      title: _string(json['title']),
      description: _string(json['description']),
      store: _string(json['store']),
      discountText: _string(json['discountText'] ?? json['discount_text']),
      code: _nullableString(json['code']),
      landingUrl: _string(json['landingUrl'] ?? json['landing_url']),
      affiliateUrl: _nullableString(
        json['affiliateUrl'] ?? json['affiliate_url'],
      ),
      imageUrl: _nullableString(json['imageUrl'] ?? json['image_url']),
      providerId: _nullableString(json['providerId'] ?? json['provider_id']),
      monetizationMode: _string(
        json['monetizationMode'] ?? json['monetization_mode'],
        fallback: 'affiliate',
      ),
      availabilityCountries: _stringList(
        json['availabilityCountries'] ?? json['availability_countries'],
      ),
      isGlobal: _bool(json['isGlobal'] ?? json['is_global']),
      validFrom: _dateTime(json['validFrom'] ?? json['valid_from']),
      validUntil: _dateTime(json['validUntil'] ?? json['valid_until']),
      featured: _bool(json['featured']),
      updatedAt:
          _dateTime(json['updatedAt'] ?? json['updated_at']) ?? DateTime.now(),
    );
  }

  static String _string(dynamic value, {String fallback = ''}) {
    if (value == null) return fallback;
    return value.toString();
  }

  static String? _nullableString(dynamic value) {
    if (value == null) return null;
    final text = value.toString().trim();
    return text.isEmpty ? null : text;
  }

  static List<String> _stringList(dynamic value) {
    if (value is! List) return const <String>[];
    return value
        .whereType<Object>()
        .map((item) => item.toString().trim())
        .where((item) => item.isNotEmpty)
        .toList(growable: false);
  }

  static bool _bool(dynamic value) {
    if (value is bool) return value;
    if (value is num) return value != 0;
    if (value is String) return value.toLowerCase() == 'true' || value == '1';
    return false;
  }

  static DateTime? _dateTime(dynamic value) {
    if (value == null) return null;
    if (value is DateTime) return value;
    final text = value.toString().trim();
    if (text.isEmpty) return null;
    return DateTime.tryParse(text);
  }
}

File: backend/app/models/promotion.py
from datetime import datetime
from typing import Literal
from pydantic import Field

from app.models.common import ApiModel
from app.models.deal import DealMonetizationMode


PromotionType = Literal[
    "coupon",
    "sale",
    "flash_sale",
]

PromotionSort = Literal[
    "featured",
    "ending_soon",
    "newest",
]


class Promotion(ApiModel):
    id: str
    type: PromotionType = "sale"
    title: str
    description: str = ""
    store: str
    discount_text: str = ""
    code: str | None = None
    landing_url: str
    affiliate_url: str | None = None
    image_url: str | None = None
    provider_id: str | None = None
    monetization_mode: DealMonetizationMode = "affiliate"
    valid_from: datetime | None = None
    valid_until: datetime | None = None
    featured: bool = False
    updated_at: datetime


class PromotionResponse(ApiModel):
    id: str
    type: PromotionType
    title: str
    description: str
    store: str
    discount_text: str
    code: str | None = None
    landing_url: str
    affiliate_url: str | None = None
    image_url: str | None = None
    provider_id: str | None = None
    monetization_mode: DealMonetizationMode = "affiliate"
    valid_from: datetime | None = None
    valid_until: datetime | None = None
    featured: bool = False
    updated_at: datetime


class PromotionsPage(ApiModel):
    items: list[PromotionResponse]
    total: int
    page: int
    page_size: int
    has_next_page: bool


class PromotionUpsertRequest(ApiModel):
    id: str
    type: PromotionType = "sale"
    title: str = Field(min_length=1)
    description: str = ""
    store: str = Field(min_length=1)
    discount_text: str = ""
    code: str | None = None
    landing_url: str = Field(min_length=1)
    affiliate_url: str | None = None
    image_url: str | None = None
    provider_id: str | None = None
    monetization_mode: DealMonetizationMode | None = None
    valid_from: datetime | None = None
    valid_until: datetime | None = None
    featured: bool = False
    updated_at: datetime | None = None


class PromotionCleanupResponse(ApiModel):
    status: str
    message: str
    checked_count: int = 0
    deleted_count: int = 0
    remaining_count: int = 0
    deleted_reasons: dict[str, int] = Field(default_factory=dict)

class AwinPromotionSyncRequest(ApiModel):
    membership: Literal["joined", "notJoined", "all"] = "joined"
    status: Literal["active", "expiringSoon", "upcoming"] = "active"
    type: Literal["promotion", "voucher", "all"] = "all"
    page_size: int = Field(default=100, ge=10, le=200)
    max_pages: int = Field(default=5, ge=1, le=50)
    advertiser_ids: list[int] | None = None
    region_codes: list[str] | None = None
    exclusive_only: bool | None = None
    updated_since: str | None = None


class AwinPromotionSyncResponse(ApiModel):
    status: str
    message: str
    fetched_count: int
    imported_count: int
    total_before: int
    total_after: int
    skipped_count: int = 0
    pages_checked: int = 0
    cleanup_deleted_count: int = 0
    cleanup_deleted_reasons: dict[str, int] = Field(default_factory=dict)
